**Abstract:**

The increasing number of workers who have diabetes in the farmer’s society, we decided to work on AI applications that can help estimate prediabetic based on data provided by Dr. Angelos Sikalidis. In the project, we also built a recommendation system based on AI to recommend food based on the number of calories, protein, and lipid given to the recommender system. We used different libraries such as NumPy, pandas, scikit-learn, and Keras library in this project.
**NOT FINISH**

In [1]:
#import required libraries
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from sklearn.model_selection import train_test_split
import numpy as np

Let's load in the CSV. This file is the mock data of different measurements taken of individual farm workers.
The measurements are:
- BMI
- Weight (Kg)
- Waist Circumfrence (cm)
- Systolic BP (mmHg)
- Diastolic BP (mmHg)
- HDL-c (mg/dl)
- LDL-c (mg/dl)
- Total Cholesterol (mg/dL)
- Atherogenicity index
- TAG (mg/dL)
- FPG (mg/dL)
- HbA1c (%)
- CRP (mg/L)
- Pulse Oximetry (%)
- Servings Grain (oz.)
- Servings Fruit (2c.)
- Servings Veg (2.5c.)
- Servings Pro (5.5oz.)
- Servings Dairy (1c.)
- Total Caloric Intake (kcal)
- % Energy CHO
- % Energy Pro
- % Energy Fat
- Vitamin D (mcg)
- Vitamin E (mg)
- Vitamin A (mcg)
- Vitamin K (mcg)
- Vitamin B12 (mcg)
- Folate (mcg)
- Category [Malnourished, Normal, Borderline Risk, Overweight At Risk, High Risk]
- Diabetic [Non-diabetic, pre-diabetic, diabetic]

First thing we want to do is classify each individuals into certain health categories. These health categories include:
- **Malnourished: suffering from malnutrition.**
- **Normal: conforming to a standard health.**
- **Borderline Risk: close to bad health.**
- **Overweight At Risk: above a weight considered normal or desirable.**
- **High Risk: very dangerous health**

In [94]:
df = pd.read_csv('mock_data.csv')
properties = list(df.columns.values)
properties.remove('Category')

y = df['Category']
c = [p for p in range(len(y))]
    
m_data = pd.DataFrame(np.zeros((len(y),len(properties))),columns=properties,index=c)

for i in range(df.shape[0]):
    for p in properties:
        data = df.iloc[i][p]
        if type(data) == str:
            print("We got a string, should not happen")
            data = np.float64(data.replace(',',''))
            if np.isnan(data):
                #outcome = df.iloc[i]['Category']
                cleanData(m_data,p,outcome,i)
            else:
                m_data.iloc[i][p] = data
                print("-Added to m_data")
        elif np.isnan(data):
            #outcome = df.iloc[i]['Category']
            pass
        else:
            m_data.iloc[i][p] = data

X = m_data[properties]
X.head

<bound method NDFrame.head of       BMI  Weight (Kg)  Waist Circumfrence (cm)  Systolic BP (mmHg)  \
0    17.0         50.0                     70.0                55.0   
1    18.0         54.0                     73.0                60.0   
2    16.0         56.0                     72.0                59.0   
3    17.0         57.0                     71.0                57.0   
4    18.0         53.0                     70.0                59.0   
..    ...          ...                      ...                 ...   
114  50.0        106.0                    105.0               182.0   
115  47.0        119.0                    106.0               181.0   
116  45.0        107.0                    107.0              1184.0   
117  47.0        118.0                    109.0               185.0   
118  50.0        120.0                    110.0               184.0   

     Diastolic BP (mmHg)  HDL-c (mg/dl)  LDL-c (mg/dl)  \
0                   66.0           27.0           55.0   
1

In [3]:
# Once we load in the file, lets save 20% of our data for testing.
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=0)
y_train = keras.utils.to_categorical(y_train) #turning targets into categorical vectors training
y_test = keras.utils.to_categorical(y_test) #turning targets into categorical vectors testing
num_classes = len(y_test[0])

feature_vector_length = len(properties)
input_shape = (feature_vector_length,)

In [8]:
# Configure the model
c_model = keras.Sequential()
c_model.add(keras.layers.Dense(32,input_shape=input_shape,activation=tf.nn.relu,kernel_initializer='he_uniform'))
c_model.add(keras.layers.Dense(32,activation=tf.nn.relu,kernel_initializer='he_uniform'))
c_model.add(keras.layers.Dense(num_classes,activation=tf.nn.softmax))

# Compile the model
c_model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
history = c_model.fit(X_train,y_train,epochs=64,verbose=1,validation_split=0.2)

Epoch 1/64
3/3 [==============================] - 0s 53ms/step - loss: 858.2748 - accuracy: 0.2632 - val_loss: 718.4388 - val_accuracy: 0.2105
Epoch 2/64
3/3 [==============================] - 0s 11ms/step - loss: 732.8979 - accuracy: 0.0526 - val_loss: 636.4836 - val_accuracy: 0.0000e+00
Epoch 3/64
3/3 [==============================] - 0s 11ms/step - loss: 627.0848 - accuracy: 0.0000e+00 - val_loss: 531.3884 - val_accuracy: 0.0526
Epoch 4/64
3/3 [==============================] - 0s 8ms/step - loss: 519.2374 - accuracy: 0.0526 - val_loss: 445.0639 - val_accuracy: 0.4737
Epoch 5/64
3/3 [==============================] - 0s 8ms/step - loss: 434.7637 - accuracy: 0.3158 - val_loss: 375.6397 - val_accuracy: 0.5263
Epoch 6/64
3/3 [==============================] - 0s 8ms/step - loss: 358.8833 - accuracy: 0.3816 - val_loss: 294.5411 - val_accuracy: 0.5263
Epoch 7/64
3/3 [==============================] - 0s 9ms/step - loss: 280.2735 - accuracy: 0.3684 - val_loss: 207.7782 - val_accuracy: 0.

3/3 [==============================] - 0s 13ms/step - loss: 0.3874 - accuracy: 0.9211 - val_loss: 1.0037 - val_accuracy: 0.8947
Epoch 60/64
3/3 [==============================] - 0s 21ms/step - loss: 0.6853 - accuracy: 0.9211 - val_loss: 0.6699 - val_accuracy: 0.8947
Epoch 61/64
3/3 [==============================] - 0s 23ms/step - loss: 0.4358 - accuracy: 0.9342 - val_loss: 0.5200 - val_accuracy: 0.8947
Epoch 62/64
3/3 [==============================] - 0s 8ms/step - loss: 0.3348 - accuracy: 0.9605 - val_loss: 0.1553 - val_accuracy: 0.9474
Epoch 63/64
3/3 [==============================] - 0s 11ms/step - loss: 0.5562 - accuracy: 0.9211 - val_loss: 0.2308 - val_accuracy: 0.8947
Epoch 64/64
3/3 [==============================] - 0s 9ms/step - loss: 0.8339 - accuracy: 0.8947 - val_loss: 0.4799 - val_accuracy: 0.8947


In [9]:
test_results = c_model.evaluate(X_test,y_test,verbose=1)
print(f'Test results - Loss: {test_results[0]} - Accuracy: {test_results[1]*100}%')

# Save the model if it is 100% accurate
if test_results[1]*100 == 100:
    c_model.save('models')

1/1 [==============================] - 0s 1ms/step - loss: 1.2481 - accuracy: 0.9167
Test results - Loss: 1.2480813264846802 - Accuracy: 91.66666865348816%


This model that we trained gives us an accuracy of 91.66%. Let's see what our model predicts for other data not used the mock data.

Our farmer has the following:

[BMI,

Weight (Kg),

Waist Circumfrence (cm),

Systolic BP (mmHg),

Diastolic BP (mmHg),

HDL-c (mg/dl),

LDL-c (mg/dl),

Total Cholesterol (mg/dL),

Atherogenicity index,

TAG (mg/dL),

FPG (mg/dL),

HbA1c (%),

CRP (mg/L),

Pulse Oximetry (%),

Servings Grain (oz.),

Servings Fruit (2c.),

Servings Veg (2.5c.),

Servings Pro (5.5oz.),

Servings Dairy (1c.),

Total Caloric Intake (kcal),

% Energy CHO,

% Energy Pro,

% Energy Fat,

Vitamin D (mcg),

Vitamin E (mg),

Vitamin A (mcg),

Vitamin K (mcg),

Vitamin B12 (mcg),

Folate (mcg),

Diabetic]

**Farmer 0: [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0,0]**

This person is not diabetic but is **malnourished**. 

**Farmer 1: [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,1,0]**

This person is not diabetic and has **normal health status**. 

**Farmer 2: [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,2,1]**

This person is not diabetic and has **borderline risk health status**. 

**Farmer 3: [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,3,2]**

This person is diabetic and is **overweight at risk health status**.

**Farmer 4: [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,4,2]**

This person is also diabetic and is **high risk health status**

In [45]:
farmer_00 = [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]
farmer_01 = [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,0]
farmer_02 = [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,1]
farmer_03 = [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,2]
farmer_04 = [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,2]

t = [farmer_00,farmer_01,farmer_02,farmer_03,farmer_04]
res = c_model.predict(t)

for index,result in enumerate(res):
    print("Farmer ",index)
    print(res[index])
    print("-----------------------------------------------------------------------")

Farmer  0
[1.0000000e+00 2.6512515e-17 4.3650728e-26 0.0000000e+00 0.0000000e+00]
-----------------------------------------------------------------------
Farmer  1
[1.5256412e-01 8.4743583e-01 1.3507615e-24 0.0000000e+00 6.0952655e-16]
-----------------------------------------------------------------------
Farmer  2
[5.5200005e-23 0.0000000e+00 1.7562917e-02 9.8243713e-01 1.8863593e-28]
-----------------------------------------------------------------------
Farmer  3
[8.0255252e-21 0.0000000e+00 1.8665767e-25 1.0000000e+00 0.0000000e+00]
-----------------------------------------------------------------------
Farmer  4
[0. 0. 0. 0. 1.]
-----------------------------------------------------------------------


The output of our model is an array of probability for each indices 0-4.

Each index is a class for a health category.
- 0: Malnourished
- 1: Normal Health Status
- 2: Boderline Risk Health Status
- 3: Overweight at Risk Health Status
- 4: High Risk Health Status

At each indices, the probabilty tells us how likely it is that the farmer falls into the above category. Looking at famer 0:

Farmer  0: [1.0000000e+00 , 2.6512515e-17 , 4.3650728e-26 , 0.0000000e+00 , 0.0000000e+00]

At index 0, we have 1.0000000e+00 which is a probability of 100%.

At index 1, we have 2.6512515e-17 which is a probability of ~0%.

At index 2, we have 4.3650728e-26 which is a probability of ~0%.

At index 3, we have 0.0000000e+00 which is a probability of 0%.

At index 4, we have 0.0000000e+00 which is a probability of 0%.

The model predicts that the farmer falls into index 0 category which is Malnourished, the correct health status.

In [37]:
for i,rr in enumerate(res):
    max_i = 0
    max_r = 0
    print("Farmer ",i)
    if i == 0:
        print("True= Malnourished")
    elif i == 1:
        print("True= Normal")
    elif i == 2:
        print("True= Borderline Risk")
    elif i == 3:
        print("True= Overweight At Risk")
    elif i == 4:
        print("True= High Risk")
    for ri,r in enumerate(rr):
        if r > max_r:
            max_i = ri
            max_r = r
    if max_i == 0:
        print("Pred= Malnourished")
    elif max_i == 1:
        print("Pred= Normal")
    elif max_i == 2:
        print("Pred= Borderline Risk")
    elif max_i == 3:
        print("Pred= Overweight At Risk")
    elif max_i == 4:
        print("Pred= High Risk")
    print("-------------------------")

Farmer  0
True= Malnourished
Pred= Malnourished
-------------------------
Farmer  1
True= Normal
Pred= Normal
-------------------------
Farmer  2
True= Borderline Risk
Pred= Overweight At Risk
-------------------------
Farmer  3
True= Overweight At Risk
Pred= Overweight At Risk
-------------------------
Farmer  4
True= High Risk
Pred= High Risk
-------------------------


From the output of our model compared to the true value of our data, for all the tests, the prediction of the model matches our data. However, if we retrain our model many times again, we get very different accuracies ranging from 60% accurate to 100% accuarate. We still cannot say that our model is 100% accurate because our data size is only 119. A much bigger  data will give us a more accutate result. 

## Prediabetic or Diabetic ##

Now that we can classify each individuals into certain health categories, it would be very useful to know if these individuals are diabetic, pre-diabetic or not diabetic. This way when we recommend food to a person, we can be sure we're not recommending diabetics a certain food they aren't allowed to consume. 

In [14]:
properties = list(df.columns.values)
properties.remove('Diabetic')

y_diabetes = df['Diabetic']

In [15]:
# Splitting test and training sets
Xdiabetes_train, Xdiabetes_test, ydiabetes_train, ydiabetes_test = train_test_split(X,y_diabetes,test_size=0.2,random_state=0)
ydiabetes_train = keras.utils.to_categorical(ydiabetes_train) #turning targets into categorical vectors training
ydiabetes_test = keras.utils.to_categorical(ydiabetes_test) #turning targets into categorical vectors testing
num_classes = len(ydiabetes_test[0])

feature_vector_length = len(properties)
input_shape = (feature_vector_length,)

In [16]:
diabetes_model = keras.Sequential()
diabetes_model.add(keras.layers.Dense(32,input_shape=input_shape,activation=tf.nn.relu,kernel_initializer='he_uniform'))
diabetes_model.add(keras.layers.Dense(32,activation=tf.nn.relu,kernel_initializer='he_uniform'))
diabetes_model.add(keras.layers.Dense(num_classes,activation=tf.nn.softmax))

diabetes_model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
history_diabetes = diabetes_model.fit(Xdiabetes_train,ydiabetes_train,epochs=64,verbose=1,validation_split=0.2)

Epoch 1/64
3/3 [==============================] - 0s 72ms/step - loss: 768.4158 - accuracy: 0.2763 - val_loss: 625.2527 - val_accuracy: 0.3684
Epoch 2/64
3/3 [==============================] - 0s 11ms/step - loss: 672.5027 - accuracy: 0.2763 - val_loss: 545.5032 - val_accuracy: 0.3684
Epoch 3/64
3/3 [==============================] - 0s 15ms/step - loss: 580.9590 - accuracy: 0.2763 - val_loss: 464.8070 - val_accuracy: 0.3684
Epoch 4/64
3/3 [==============================] - 0s 18ms/step - loss: 490.2589 - accuracy: 0.2763 - val_loss: 386.3090 - val_accuracy: 0.3684
Epoch 5/64
3/3 [==============================] - 0s 24ms/step - loss: 406.2060 - accuracy: 0.2763 - val_loss: 307.6856 - val_accuracy: 0.3684
Epoch 6/64
3/3 [==============================] - 0s 18ms/step - loss: 318.7429 - accuracy: 0.2763 - val_loss: 232.4138 - val_accuracy: 0.3684
Epoch 7/64
3/3 [==============================] - 0s 10ms/step - loss: 238.8600 - accuracy: 0.2763 - val_loss: 160.0174 - val_accuracy: 0.3684

3/3 [==============================] - 0s 11ms/step - loss: 1.8111 - accuracy: 0.8553 - val_loss: 1.3027 - val_accuracy: 0.8421
Epoch 60/64
3/3 [==============================] - 0s 13ms/step - loss: 1.6196 - accuracy: 0.8684 - val_loss: 0.8199 - val_accuracy: 0.7895
Epoch 61/64
3/3 [==============================] - 0s 12ms/step - loss: 1.8750 - accuracy: 0.9211 - val_loss: 2.0597 - val_accuracy: 0.8421
Epoch 62/64
3/3 [==============================] - 0s 9ms/step - loss: 2.0011 - accuracy: 0.8158 - val_loss: 2.2407 - val_accuracy: 0.7895
Epoch 63/64
3/3 [==============================] - 0s 10ms/step - loss: 2.4002 - accuracy: 0.7368 - val_loss: 1.6525 - val_accuracy: 0.8421
Epoch 64/64
3/3 [==============================] - 0s 10ms/step - loss: 1.1909 - accuracy: 0.8947 - val_loss: 1.7756 - val_accuracy: 0.8947


In [28]:
test_results_diabetes = diabetes_model.evaluate(Xdiabetes_test,ydiabetes_test,verbose=1)
print(f'Test results - Loss: {test_results_diabetes[0]} - Accuracy: {test_results_diabetes[1]*100}%')

# Save the model if it is 100% accurate
if test_results_diabetes[1]*100 == 100:
    diabetes_model.save('models')

1/1 [==============================] - 0s 1ms/step - loss: 0.0578 - accuracy: 0.9583
Test results - Loss: 0.05775165557861328 - Accuracy: 95.83333134651184%


Using the same farmers as before, lets see what our model predicts.

**Farmer 0: [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]**

This person is **not diabetic** but is malnourished

**Farmer 1: [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,1]**

This person is **not diabetic** and has normal health status.

**Farmer 2: [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,2]**

This person is **not diabetic** and has borderline risk health status

**Farmer 3: [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,3]**

This person is **diabetic** and is overweight at risk health status

**Farmer 4: [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,4]**

This person is also **diabetic** and is high risk health status


In [35]:
farmer_00 = [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]
farmer_01 = [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,1]
farmer_02 = [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,2]
farmer_03 = [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,3]
farmer_04 = [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,4]

res_diabetic = diabetes_model.predict([farmer_00,farmer_01,farmer_02,farmer_03,farmer_04])
#print(res_diabetic)

for i,result in enumerate(res_diabetic):
    print("Farmer ",i)
    max_r = 0
    max_i = 0
    for ir,r in enumerate(result):
        if r > max_r:
            max_r = r
            max_i = ir
    if i == 0:
        print("True: Not diabetic")
    elif i == 1:
        print("True: Not diabetic")
    elif i == 2:
        print("True: Not diabetic")
    elif i == 3:
        print("True: Diabetic")
    elif i == 4:
        print("True: Diabetic")
    if max_i == 0:
        print("Pred: Not diabetic")
    elif max_i == 1:
        print("Pred: Prediabetic")
    else:
        print("Pred: Diabetic")
    print("------------------")


Farmer  0
True: Not diabetic
Pred: Not diabetic
------------------
Farmer  1
True: Not diabetic
Pred: Not diabetic
------------------
Farmer  2
True: Not diabetic
Pred: Prediabetic
------------------
Farmer  3
True: Diabetic
Pred: Diabetic
------------------
Farmer  4
True: Diabetic
Pred: Diabetic
------------------


## Recommendation System

**A very basaic CSV file that contains about 7,412 food variety and their nutritional facts**

In [53]:
#import external libraries and functions
import sklearn
from sklearn.neighbors import NearestNeighbors

#load dataset from CSV file and show first 5 records
food_bank = pd.read_csv('food.csv')
   
food_bank.columns = food_bank.columns
food_bank.head()

Category               Description  Nutrient Data Bank Number  \
0      BUTTER          BUTTER,WITH SALT                       1001   
1      BUTTER  BUTTER,WHIPPED,WITH SALT                       1002   
2  BUTTER OIL      BUTTER OIL,ANHYDROUS                       1003   
3      CHEESE               CHEESE,BLUE                       1004   
4      CHEESE              CHEESE,BRICK                       1005   

   Data.Alpha Carotene  Data.Ash  Data.Beta Carotene  Data.Beta Cryptoxanthin  \
0                    0      2.11                 158                        0   
1                    0      2.11                 158                        0   
2                    0      0.00                 193                        0   
3                    0      5.11                  74                        0   
4                    0      3.18                  76                        0   

   Data.Carbohydrate  Data.Cholesterol  Data.Choline  ...  \
0               0.06               215            19  ...   
1               0.06               219            19  ...   
2               0.00               256            22  ...   
3               2.34                75            15  ...   
4               2.79                94            15  ...   

   Data.Major Minerals.Potassium  Data.Major Minerals.Sodium  \
0                             24                         576   
1                             26                         827   
2                              5                           2   
3                            256                        1395   
4                            136                         560   

   Data.Major Minerals.Zinc  Data.Vitamins.Vitamin A - IU  \
0                      0.09                          2499   
1                      0.05                          2499   
2                      0.01                          3069   
3                      2.66                           763   
4                      2.60                          1080   

   Data.Vitamins.Vitamin A - RAE  Data.Vitamins.Vitamin B12  \
0                            684                       0.17   
1                            684                       0.13   
2                            840                       0.01   
3                            198                       1.22   
4                            292                       1.26   

   Data.Vitamins.Vitamin B6  Data.Vitamins.Vitamin C  Data.Vitamins.Vitamin E  \
0                     0.003                      0.0                     2.32   
1                     0.003                      0.0                     2.32   
2                     0.001                      0.0                     2.80   
3                     0.166                      0.0                     0.25   
4                     0.065                      0.0                     0.26   

   Data.Vitamins.Vitamin K  
0                      7.0  
1                      7.0  
2                      8.6  
3                      2.4  
4                      2.5  

[5 rows x 48 columns]

In [54]:
data_key = pd.read_csv('mock_data_key.csv')

print(data_key)

                            MALNOURISHED NORMAL/DESIRABLE BORDERLINE RISK  \
BMI                                 < 19            19-25           26-29   
Weight (Kg)                         < 63            63-68           69-86   
Waist Circumference (cm)            ≤ 73            74-81           82-89   
Systolic BP (mmHg)                  ≤ 60           61-129         130-139   
Diastolic BP (mmHg)                 ≤ 67            68-79           80-89   
HDL-c (mg/dL)                       < 30           60-100           46-59   
LDL-c (mg/dL)                       < 65            65-90          91-130   
Total Cholesterol (mg/dL)          < 110          160-180         181-200   
Atherogenicity index                   X         −0.3-0.1       0.11-0.24   
TAG (mg/dL)                            X            < 150         150-200   
FPG (mg/dL)                         < 75           75-100         100-115   
HbA1c (%)                            < 4            4-5.7         5.7-6.4   

**The data_key tells us where each measurements of our features needs to be at in order to fall under certain health categories.**

Lets try to recommend a food to this person: 

farmer_00 = [17,
            50,
            70,
            55,
            66,
            27,
            55,
            100,
            0,
            0,
            48,
            3.8,
            0,
            84,
            2,
            0,
            0,
            1,
            0,
            1799,
            40,
            20,
            15,
            25,
            **10 Vitamin E,
            699 Vitamin A,
            75 Vitamin K,
            1.4 Vitamin B12,**
            390,
            0]

This famer is categorized as malnourished and has 
- 10 mg of Vitamin E (mg)
- 699 mcg of Vitamin A (mcg)
- 75 mcg of Vitamin K (mcg)
- 1.4 mcg of Vitamin B12 (mcg)

The recommended amount:
- Vitamin E is 12-18 mg
- Vitamin A is 700-1000 mcg
- Vitamin K is 110-140  mcg
- Vitamin B12 is 400-600 mcg

We can find the difference of the recommended and actual to get the proper food recommendation.

Need:
- 2 mg of Vitamin E
- 1 mcg of Vitamin A
- 35 mcg of Vitamin K
- 398.2 mcg of Vitamin B12

Since we lack the most on Vitamin K and Vitamin B12, we need to recommend a food that will help normalize these levels.

In [155]:
farmer_00 = [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]

# Indices to where each farmer's vitamin measures are
v_e = 24
v_a = 25
v_k = 26
v_b12 = 27

# Food bank indices of vitamins we want
f_a_index = 41
f_b12_index = 43
f_e_index = 46
f_k_index = 47

# Recommended Vitamins
rec_vitamins = {}
rec_vitamins["vitamin_e"] = 12
rec_vitamins["vitamin_a"] = 700
rec_vitamins["vitamin_k"] = 110
rec_vitamins["vitamin_b12"] = 400

recommendations = []

def get_food_rec(farmer):
    need_vitamin_e = rec_vitamins["vitamin_e"] - farmer[v_e]
    need_vitamin_a = rec_vitamins["vitamin_a"] - farmer[v_a]
    need_vitamin_k = rec_vitamins["vitamin_k"] - farmer[v_k]
    need_vitamin_b12 = rec_vitamins["vitamin_b12"] - farmer[v_b12]
    
    need_list = []
    need_list.append(need_vitamin_e)
    need_list.append(need_vitamin_a)
    need_list.append(need_vitamin_k)
    need_list.append(need_vitamin_b12)
    
    foods = {}
    
    vitamins_needed = {}
    vitamin = "Vitamin_E"
    vitamins_needed[vitamin] = 0
    vitamin = "Vitamin_A"
    vitamins_needed[vitamin] = 0
    vitamin = "Vitamin_K"
    vitamins_needed[vitamin] = 0
    vitamin = "Vitamin_B12"
    vitamins_needed[vitamin] = 0

    for ni,n in enumerate(need_list):
        if n > 5:
            vitamin = ""
            if ni == 0:
                vitamin = "Vitamin_E"
                vitamins_needed[vitamin] = n
            elif ni == 1:
                vitamin = "Vitamin_A"
                vitamins_needed[vitamin] = n
            elif ni == 2:
                vitamin = "Vitamin_K"
                vitamins_needed[vitamin] = n
            elif ni == 3:
                vitamin = "Vitamin_B12"
                vitamins_needed[vitamin] = n

    X_rec = food_bank.iloc[:,[f_e_index,f_a_index,f_k_index,f_b12_index]].values
    nbrs = NearestNeighbors(n_neighbors=1).fit(X_rec)
    
    not_in_range = True
    for key in vitamins_needed:
        if vitamins_needed[key] != 0:
            if key == "Vitamin_E":
                food = food_bank.iloc[nbrs.kneighbors([[vitamins_needed["Vitamin_E"],0,0,0]])[1][0][0]]
                foods["Vitamin_E"] = food
                
            elif key == "Vitamin_A":
                food = food_bank.iloc[nbrs.kneighbors([[0,vitamins_needed["Vitamin_A"],0,0]])[1][0][0]]
                foods["Vitamin_A"] = food
            elif key == "Vitamin_K":
                food = food_bank.iloc[nbrs.kneighbors([[0,0,vitamins_needed["Vitamin_K"],0]])[1][0][0]]
                foods["Vitamin_K"] = food
            elif key == "Vitamin_B12":
                food = food_bank.iloc[nbrs.kneighbors([[0,0,0,vitamins_needed["Vitamin_B12"]]])[1][0][0]]
                foods["Vitamin_B12"] = food
            
    for fkey in foods:
        print("For " + fkey + " we recommend:",foods[fkey][1])
        print(foods[fkey])
        print("-------------------------------------------------------------------------------------------")

get_food_rec(farmer_00)

For Vitamin_K we recommend: CASHEW NUTS,OIL RSTD,W/SALT
Category                                                                   CASHEW NUTS
Description                                                CASHEW NUTS,OIL RSTD,W/SALT
Nutrient Data Bank Number                                                        12586
Data.Alpha Carotene                                                                  0
Data.Ash                                                                          2.89
Data.Beta Carotene                                                                   0
Data.Beta Cryptoxanthin                                                              0
Data.Carbohydrate                                                                30.16
Data.Cholesterol                                                                     0
Data.Choline                                                                        61
Data.Fiber                                                                

We got these 2 foods: **CASHEW NUTS (Vitamin K)** and **OYSTER (Vitamin B12)**

For Cashew Nuts, we needed about 35 mcg more.

The information about it on the Food CVS gives us the following:

Data.Vitamins.Vitamin A:                                                         0

Data.Vitamins.Vitamin B12:                                                            0

Data.Vitamins.Vitamin E:                                                           0.92

**Data.Vitamins.Vitamin K:                                                           34.7**

For oysters, we needed about 398 mcg more. We have the following about oysters:

Data.Vitamins.Vitamin A:                                                             0

**Data.Vitamins.Vitamin B12:                                                              27.8**

Data.Vitamins.Vitamin E:                                                                   0

Data.Vitamins.Vitamin K:                                                                   0